### PR TITLE
A naive implement trainer.train by executor

### PR DESCRIPTION
**Pull request by reyoung**
This PR is based on #10303 and implement the trainer.train very naively by the executor.

Related issue #10344

### CODE REVIEW

File: python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+                              act='sigmoid')
+    predict_word = fluid.layers.fc(input=hidden1, size=dict_size, act='softmax')
+    return predict_word
+
+
+def train_network(is_sparse):
+    next_word = fluid.layers.data(name='nextw', shape=[1], dtype='int64')
+    predict_word = inference_network(is_sparse)
+    cost = fluid.layers.cross_entropy(input=predict_word, label=next_word)
+    avg_cost = fluid.layers.mean(cost)
```
Comment (reyoung) [inline python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py]: The metrics, like error rate, auc, are not defined in this demo. How do we define a error rate in our new designed API?

By just adding `fluid.layers.accury(input=predict_word, label=label)` ? Should we fetch metrics by default?

File: python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+def train(use_cuda, is_sparse, save_path):
+    train_reader = paddle.batch(
+        paddle.dataset.imikolov.train(word_dict, N), BATCH_SIZE)
+
+    place = fluid.CUDAPlace(0) if use_cuda else fluid.CPUPlace()
+
+    def event_handler(event):
+        print type(event)
+        if isinstance(event, fluid.EndEpochEvent):
+            avg_cost = trainer.test(reader=paddle.dataset.imikolov.test(
```
Comment (JiayiFeng) [inline python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py]: The constructor of `trainer` only takes `train_net`. The `trainer` doesn't know which part of the net should be used for testing at all. How to implement `trainer.test()` ?

File: python/paddle/fluid/__init__.py [hunk 3]
```diff
@@ -21,8 +21,7 @@
 from executor import *
 
 import trainer
-from trainer import Trainer
-from trainer import Event
+from trainer import *
```
Comment (wangkuiyi) [inline python/paddle/fluid/__init__.py]: No `import *` please, due to https://shahriar.svbtle.com/importing-star-in-python

File: python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    res.set(data, place)
+    res.set_lod([lod])
+    return res
+
+
+word_dict = paddle.dataset.imikolov.build_dict()
+dict_size = len(word_dict)
+
+
+def inference_network(is_sparse):
```
Comment (wangkuiyi) [inline python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py]: Let us rename `inference_network` into `inference_program`, and do not use the misleading term *network* with Fluid anyway.

File: python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    concat_embed = fluid.layers.concat(
+        input=[embed_first, embed_second, embed_third, embed_forth], axis=1)
+    hidden1 = fluid.layers.fc(input=concat_embed,
+                              size=HIDDEN_SIZE,
+                              act='sigmoid')
+    predict_word = fluid.layers.fc(input=hidden1, size=dict_size, act='softmax')
+    return predict_word
+
+
+def train_network(is_sparse):
```
Comment (wangkuiyi) [inline python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py]: Similar to the above comment, please rename `train_network` into `train_program`.

File: python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py (hunk 2 above)
Comment (wangkuiyi) [reply python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py]: I had the same question and I asked @helinwang . @helinwang explained his idea 

- The `Trainer.train` method runs `train_network`, which should be renamed into `train_program`, twice
  1. At the first time, it generates a *test startup ProgramDesc* and a *test main ProgramDesc`, which will be used by the event handler to compute the `avg_cost`.  Please be aware that the two test ProgramDescs runs till the `avg_cost` and do not contain the backward pass.
  1. At the second time, the `train_program` works with the optimizer to generate the *train startup ProgramDesc* and the *train main ProgramDesc*, which include the backward pass.

So, in short, `Trainer.train` creates four `ProgramDesc` messages by running `train_program` twice:
1. test startup ProgramDesc
1. test main ProgramDesc
1. train startup ProgramDesc
1. train main ProgramDesc

Comment (helinwang) [reply python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py]: We can safely discard `test startup ProgramDesc` since its operators is a subset of the operators of `train startup ProgramDesc`.

File: python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py (hunk 1 above)
Comment (helinwang) [reply python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py]: Maybe we can let the `train_network` function can return one or more variables, the first one is always the loss variable to be minimized. The rest are the additional ones to be included in the "end step" event. E.g.,

```Python
def train_network(is_sparse):
  #...
  accuracy = fluid.layers.accury(input=predict_word, label=label)
  return avg_cost, accuracy

def event_handler(event):
  loss, accuracy = event.values
  print(loss, accuracy)
```

File: python/paddle/fluid/__init__.py (hunk 3 above)
Comment (helinwang) [reply python/paddle/fluid/__init__.py]: Fixed in https://github.com/PaddlePaddle/Paddle/pull/10354

File: python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py (hunk 4 above)
Comment (helinwang) [reply python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py]: Fixed in https://github.com/PaddlePaddle/Paddle/pull/10354

File: python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py (hunk 5 above)
Comment (helinwang) [reply python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py]: Fixed in https://github.com/PaddlePaddle/Paddle/pull/10354

File: python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py (hunk 1 above)
Comment (helinwang) [reply python/paddle/fluid/tests/book/word2vec/no_test_word2vec_new_api.py]: @cs2be 's PR https://github.com/PaddlePaddle/Paddle/pull/10339 supports it.
